django/db/models/sql/compiler.py

```diff
@@ -123,8 +123,8 @@ def get_group_by(self, select, order_by):
         for expr, (sql, params, is_ref) in order_by:
             # Skip References to the select clause, as all expressions in the
             # select clause are already part of the group by.
-            if not expr.contains_aggregate and not is_ref:
-                expressions.extend(expr.get_source_expressions())
+            if not is_ref:
+                expressions.extend(expr.get_group_by_cols())
```

Comment:
That makes sense, using `get_source_expressions` here was only performing a `contains_aggregate` check for the other level expression while `get_group_by_cols` does it recursively as the expression tree is walked.

https://github.com/django/django/blob/c1b24718e05ea474955777d7bc4d9d5634560cd5/django/db/models/expressions.py#L346-L352

Reply:
@felixxm makes me wonder if we want to do `source.get_group_by_cols(alias)` in `BaseExpression.get_group_by_cols`. The current implementation makes me think we might be suffering from an issue similar to [#31094](https://github.com/django/django/commit/5a4d7285bd10bd40d9f7e574a7c421eb21094858) if `Coalesce(Subquery)` is used instead. I'll submit a PR if it's the case.

Reply:
@charettes Thanks for checking :+1: Agreed, it looks suspicious, however I wasn't able to reproduce a similar issue (at least for now). I can play with this a bit more next week.